consensus/chain_selector.go

```diff
@@ -33,8 +48,62 @@ func NewChainSelector(cs *hamt.CborIpldStore, actorState SnapshotGenerator, gCid
 	}
 }
 
-// Weight returns the expected consensus weight of this TipSet in uint64
-// encoded fixed point representation.
+// NewWeight returns the EC weight of this TipSet in uint64 encoded fixed point
+// representation.
+//
+// w(i) = w(i-1) + (PI)^(P_n) * [V * num_blks + X ]
+// P_n(n) = if n < 3:0 else: n, n is number of null rounds
+// X = log_2(total_storage(pSt))
+func (c *ChainSelector) NewWeight(ctx context.Context, ts types.TipSet, pSt state.Tree) (uint64, error) {
+	ctx = log.Start(ctx, "Expected.Weight")
+	log.LogKV(ctx, "Weight", ts.String())
+	if ts.Len() == 1 && ts.At(0).Cid().Equals(c.genesisCid) {
+		return uint64(0), nil
+	}
+	// Compute parent weight.
+	parentW, err := ts.ParentWeight()
+	if err != nil {
+		return uint64(0), err
+	}
+
+	w, err := types.FixedToBig(parentW)
+	if err != nil {
+		return uint64(0), err
+	}
+
+	// Each block adds ECV to the weight's inner term
+	innerTerm := new(big.Float)
+	floatECV := new(big.Float).SetInt64(int64(NewECV))
+	floatNumBlocks := new(big.Float).SetInt64(int64(ts.Len()))
+	innerTerm.Mul(floatECV, floatNumBlocks)
+
+	// Add X to the weight's inner term
+	powerTableView := c.createPowerTableView(pSt)
+	totalBytes, err := powerTableView.Total(ctx)
+	if err != nil {
+		return uint64(0), err
+	}
+	roughLogTotalBytes := new(big.Float).SetInt64(int64(totalBytes.BigInt().BitLen()))
+	innerTerm.Add(innerTerm, roughLogTotalBytes)
+
+	// Attenuate weight by the number of tickets
+	numTickets := len(ts.At(0).Tickets)
+	P := new(big.Float).SetInt64(int64(1))
+	if numTickets >= NullThresh {
+		bigPI := new(big.Float).SetFloat64(PI)
+		// P = PI^numNull
+		for i := 0; i < numTickets; i++ {
+			P.Mul(P, bigPI)
```

Comment:
Your assumption that numTickets is small enough for this for loop to be reasonable also implies that big.NewFloat(math.Pow(PI, float64(numTickets))) should be sufficient for this calculation. Does this spec specify this algorithm?

Reply:
ah didn't even notice the power function.  When we move to the actual stable speced weight function I'll remove this sloppiness.